File: 341.flatten-nested-list-iterator.java
/*
 * @lc app=leetcode id=341 lang=java
 *
 * [341] Flatten Nested List Iterator
 */
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * public interface NestedInteger {
 *
 *     // @return true if this NestedInteger holds a single integer, rather than a nested list.
 *     public boolean isInteger();
 *
 *     // @return the single integer that this NestedInteger holds, if it holds a single integer
 *     // Return null if this NestedInteger holds a nested list
 *     public Integer getInteger();
 *
 *     // @return the nested list that this NestedInteger holds, if it holds a nested list
 *     // Return null if this NestedInteger holds a single integer
 *     public List<NestedInteger> getList();
 * }
 */
public class NestedIterator implements Iterator<Integer> {

    Stack<NestedInteger> stack = new Stack<>();

    public NestedIterator(List<NestedInteger> nestedList) {
        if (nestedList != null) {
            pushListItemsToStack(nestedList);
        }
    }

    @Override
    public Integer next() {
        while (!stack.empty()) {
            NestedInteger item = stack.pop();
            if (item.isInteger()) return item.getInteger();
            pushListItemsToStack(item.getList());
        }
        return null;
    }

    @Override
    public boolean hasNext() {
        while (!stack.empty()) {
            if (stack.peek().isInteger()) return true;
            pushListItemsToStack(stack.pop().getList());
        }
        return false;
    }

    private void pushListItemsToStack(List<NestedInteger> nestedList) {
        Stack<NestedInteger> tmp = new Stack<>();
        for (NestedInteger e : nestedList) tmp.push(e);
        while (!tmp.empty()) stack.push(tmp.pop());
    }
}

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i = new NestedIterator(nestedList);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: go/341.flatten-nested-list-iterator.go
/*
 * @lc app=leetcode id=341 lang=golang
 *
 * [341] Flatten Nested List Iterator
 */

// @lc code=start
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * type NestedInteger struct {
 * }
 *
 * // Return true if this NestedInteger holds a single integer, rather than a nested list.
 * func (this NestedInteger) IsInteger() bool {}
 *
 * // Return the single integer that this NestedInteger holds, if it holds a single integer
 * // The result is undefined if this NestedInteger holds a nested list
 * // So before calling this method, you should have a check
 * func (this NestedInteger) GetInteger() int {}
 *
 * // Set this NestedInteger to hold a single integer.
 * func (n *NestedInteger) SetInteger(value int) {}
 *
 * // Set this NestedInteger to hold a nested list and adds a nested integer to it.
 * func (this *NestedInteger) Add(elem NestedInteger) {}
 *
 * // Return the nested list that this NestedInteger holds, if it holds a nested list
 * // The list length is zero if this NestedInteger holds a single integer
 * // You can access NestedInteger's List element directly if you want to modify it
 * func (this NestedInteger) GetList() []*NestedInteger {}
 */

type NestedIterator struct {
	List []int
	Idx int
}

func Constructor(nestedList []*NestedInteger) *NestedIterator {
	return &NestedIterator{build(nestedList), 0}
}

func build(nestedList []*NestedInteger) []int {
	res := make([]int, 0)
	for _, elem := range nestedList {
		if elem.IsInteger() {
			res = append(res, elem.GetInteger())
		} else {
			res = append(res, build(elem.GetList())...)
		}
	}
	return res
}

func (this *NestedIterator) Next() int {
	val := this.List[this.Idx]
	this.Idx = this.Idx+1
	return val
}

func (this *NestedIterator) HasNext() bool {
    return this.Idx < len(this.List)
}
// @lc code=end

